src/osdatahub/DownloadsAPI/downloads_api.py

```diff
@@ -0,0 +1,158 @@
+import functools
+import logging
+import os
+from abc import ABC, abstractmethod
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from multiprocessing import cpu_count
+from pathlib import Path
+from typing import Union
+
+import requests
+from tqdm import tqdm
+
+
+class _DownloadObj:
+    """ Helper class to download a file from Downloads API
+
+    Args:
+        url (str): Direct url for the download file
+        file_name (str): Name of the file to be downloaded
+    """
+
+    def __init__(self, url: str, file_name: str, size: int):
+        self.url = url
+        self.file_name = file_name
+        self.size = size
+
+    def download(self, output_dir: Union[str, Path], overwrite: bool = False, pbar: tqdm = None) -> str:
+        """
+        Downloads file to given directory
+
+        Args:
+            output_dir (Union[str, Path]): Directory to save the downloaded file
+            overwrite (bool, optional): whether to overwrite an existing file. Defaults to False
+            pbar (tqdm, optional): tqdm progress bar to update in the event of downloading multiple files at once
+        """
+        output_path = os.path.join(output_dir, self.file_name)
+
+        if os.path.isfile(output_path) and not overwrite:
+            logging.warning(f"Overwrite is set to False and there is a file already in the location {output_path}. "
+                            f"Skipping download...")
+            return output_path
+
+        r = requests.get(self.url, stream=True)
+        size = int(r.headers.get('content-length'))
+        chunk_size = 1024
+        if r.status_code == 200:
+            with open(output_path, 'wb') as f:
+                if not pbar:
+                    pbar = tqdm(total=size, desc=self.file_name, unit="B", unit_scale=True, leave=True)
+                    for chunk in r.iter_content(chunk_size=chunk_size):
+                        f.write(chunk)
+                        f.flush()
+                        pbar.update(chunk_size)
+
+        pbar.write(f"Finished downloading {self.file_name} to {output_path}")
+        return output_path
+
+
+class _DownloadsAPIBase(ABC):
+    """Parent class for Product and DataPackage classes as part of the DownloadsAPI
+    (https://osdatahub.os.uk/docs/downloads/overview)
+
+    Args:
+        product_id (str): Valid ID for a Downloads API Product or DataPackage.
+
+    """
+    _ENDPOINT = f"https://api.os.uk/downloads/v1/"
+
+    def __init__(self, product_id: str):
+        self.id = product_id
+
+    def _endpoint(self, api_name: str) -> str:
+        """
+        Adds the api to the base endpoint
+
+        Args:
+            api_name (str): API endpoint to add to _ENDPOINT attribute
+
+        Returns:
+            str: concatenated endpoint consisting of _ENDPOINT and api_name
+        """
+        return f"{self._ENDPOINT}/{api_name}"
+
+    @functools.cached_property
```

Comment:
Not sure if these should actually be cached properties if the user is able to change the id - you could make the id and endpoint read only or private if it's not supposed to be changed after init?